### PR TITLE
PyTorch File Format API

### PR DESCRIPTION
**Pull request by jamesr66a**
This is a follow-up to https://github.com/pytorch/pytorch/pull/9794 that contains only the serialization library and exposes a cleaner API. This should later be incorporated into the module export code

### CODE REVIEW

File: torch/csrc/jit/serialization.h [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    if (file_size % kFieldAlignment != 0) {
+      throw std::runtime_error("File length is not a multiple of the alignment"
+                               " size. Is this a valid PyTorch file?");
+    }
+    readAndValidateFileFooter();
+  }
+  std::string getLastRecord() {
+    return getRecordWithKey(last_record_offset);
+  }
+  std::string getRecordWithKey(uint64_t key) {
```
Comment (zdevito) [inline torch/csrc/jit/serialization.h]: std::string will make it impossible to change this to an mmap-based implementation later. If it returned `std::shared_ptr<void>` then we  would be able to change it to one later.

File: torch/csrc/jit/serialization.h [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    if (key > file_size) {
```
Comment (zdevito) [inline torch/csrc/jit/serialization.h]: key + 64bytes > file_size, otherwise the reeds below might fail.

File: torch/csrc/jit/serialization.h [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    // TODO endian swap on platforms that need it?
+    if (!std::fwrite(&value, 8u, 1u, fp)) {
+      wrapPErrorAndThrow("Unable to write to file!");
+    }
+    cursor += 8u;
+  }
+
+  void writePad(const size_t num_bytes) {
+    uint8_t pad_val = kPadValue;
+    // TODO is there a more efficient way to do this?
```
Comment (zdevito) [inline torch/csrc/jit/serialization.h]:
```
static std::vector<char> zeros('\0', 64);
std::fwrite(&zeros.begin()...
```

**Review by zdevito — APPROVED** [review]
This looks good. Is there a reason we cannot return std::shared_ptr<void>?

File: torch/csrc/jit/serialization.h (hunk 1 above)
Comment (jamesr66a) [reply torch/csrc/jit/serialization.h]: I changed this while fucking around with pybind, i'll change it back

File: torch/csrc/jit/serialization.h (hunk 3 above)
Comment (jamesr66a) [reply torch/csrc/jit/serialization.h]: butbutbut i want the kernel to do that for me :'(

**Review by dzhulgakov — APPROVED** [review]
Looks legit :)

**Pull request closed without merging**